FORGOT PASSWORD

TarikVelic opened the pull request:
MAIN
 * Added forgot password function --------------------------------------------------------------------
 OTHER
 * <other msg> --------------------------------------------------------------------

TarikVelic committed:
FORGOT PASSWORD

--------------------------------------------------------------------
 MAIN
 * Added forgot password function
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on controllers/forgotPassword.js:
```diff
@@ -0,0 +1,17 @@
+  try {
+
+    const token = await forgotPasswordService.forgotPassword(email);
+    console.log(token);
```
remove console log

becarselman commented on controllers/forgotPassword.js:
```diff
@@ -0,0 +1,17 @@
+    console.log(token);
+    res.status(200).json({ token });
+  } catch (error) {
+    console.error('Failed to send forgot password email:', error);
```
same

becarselman commented on controllers/sendemail.js:
```diff
@@ -1,11 +1,12 @@
 const sendMailService = require('../services/sendemail')
 const errors = require("../configuration/errors")
+process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
```
what is this

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
+
+  const templateName = "forgot-password";
+  const url = `${process.env.FRONTEND_URL}/reset-password/${token}`;
```
we have file for importing env vars

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+  const token = uuidv4();
+  
+  user.resetPasswordToken = token;
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
```
you can do this on token creation.

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+    throw new Error("User not found");
+  }
+
+  const token = uuidv4();
```
token need to be jwt.

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+const sendMailService = require("./sendemail");
+
+async function forgotPassword(email) {
+  const user = await User.findOne({ email: email });
```
Move this to repository

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+  };
+
+  await sendMailService.sendEmail(email, templateName, context);
+  console.log('Forgot password email sent successfully.');
```
remove console.log

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Added jwt token
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,31 @@
+    throw new Error("User not found");
+  }
+
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
```
we have file for importing env vars.

becarselman commented on package.json:
```diff
@@ -22,15 +22,17 @@
     "nodemailer": "^6.9.1",
     "nodemailer-express-handlebars": "^6.1.0",
     "nodeman": "^1.1.2",
+    "uuid": "^9.0.0",
```
If we don't use uuid anymore we can remove it

becarselman commented on package.json:
```diff
@@ -22,15 +22,17 @@
     "joi-password-complexity": "^5.1.0",
     "jsonwebtoken": "^9.0.0",
-    "mongodb": "^5.2.0",
+    "mongodb": "^5.4.0",
```
why are you update version of this package

becarselman commented on package.json:
```diff
@@ -22,15 +22,17 @@
     "bcrypt": "^5.1.0",
     "cors": "^2.8.5",
     "dotenv": "^16.0.3",
+    "ejs": "^3.1.9",
```
If we dont use ejs remove it

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Fixing things from review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on repositories/user.js:
```diff
@@ -1,17 +1,25 @@
+}
 
-    return await User.create({email: email, password: password, role: role})
+async function setResetPasswordToken(user) {
```
Why you doing this in repository. Repository only need to communicate with db. Move it in some service file

becarselman commented on repositories/user.js:
```diff
@@ -1,17 +1,25 @@
+async function setResetPasswordToken(user) {
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
+  user.resetPasswordToken = token;
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
```
why you set expires date when we have already in jwt.

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,28 @@
+  const token = await userRepository.setResetPasswordToken(user);
+
+  const templateName = "forgot-password";
+  const resetPasswordUrl = process.env.RESET_PASSWORD_URL;
```
move this process.env in configuration file

becarselman commented on services/sendemail.js:
```diff
@@ -5,12 +5,15 @@ const fs = require("fs");
 
 require('dotenv').config()
```
I know this is not Your code but remove this line because we don't need

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Added style for password reset button
*  Fixed things from the review

--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on controllers/forgotPassword.js:
```diff
@@ -0,0 +1,15 @@
+  const { email } = req.body;
+
+  try {
+
```
Please pay attention to formatting in future

becarselman commented on repositories/user.js:
```diff
@@ -1,17 +1,19 @@
-const User = require("../db/models/User")
-
+const User = require("../db/models/User");
+const jwt = require('jsonwebtoken');
```
I think this is necessary import.

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,35 @@
+const userRepository = require("../repositories/user");
+const env = require("../configuration/env")
+
+async function setResetPasswordToken(user) {
```
same block of code we have in login service. Please create some helper function that will do this part(create token).

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,35 @@
+  const user = await userRepository.getUserByEmail(email);
+
+  if (!user) {
+    throw new Error("User not found");
```
Why you are not reuse error message from errors file. I wrote this already.

becarselman commented on configuration/env.js:
```diff
@@ -10,5 +10,6 @@ module.exports = {
     SMTP_USER: process.env.SMTP_USER,
-    SMTP_PASS: process.env.SMTP_PASS
+    SMTP_PASS: process.env.SMTP_PASS,
+    RESET_PASSWORD_URL: `${values.FRONTEND_URL}/reset-password/`
```
You don't need to keep this here. Keep just frontend url. And i don't see here import of FRONTEND_URL. Did you test it. Is this work

becarselman reviewed (changes requested):
I don't know how this working. You get email with resetPassword link. And I don't see where you create reset password function

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Review changes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Added Reset password functionality
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on controllers/resetPassword.js:
```diff
@@ -0,0 +1,15 @@
+exports.showResetPasswordForm = resetPasswordService.showResetPasswordForm;
+
+exports.resetPassword = async (req, res) => {
+  const { token } = req.params;
```
we should get token from header authorization

becarselman commented on email-templates/password-reset.success.hbs:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
```
where we use this template

becarselman commented on db/models/Token.js:
```diff
@@ -0,0 +1,23 @@
+
+const tokenSchema = new mongoose.Schema({
+  _userId: {
+    type: mongoose.Schema.Types.ObjectId,
```
we don't need to store token in db

becarselman commented on repositories/user.js:
```diff
@@ -1,17 +1,27 @@
-const User = require("../db/models/User")
-
+const User = require("../db/models/User");
+const jwt = require('jsonwebtoken');
```
as far as I can see you resolve comment and do nothing 
<img width="1078" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-backend/assets/102960105/94a4572f-5a5c-4e50-8fbc-84e641e8c581">

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,42 @@
+const userRepository = require('../repositories/user');
+const errors = require('../configuration/errors');
+
+async function showResetPasswordForm(req, res) {
```
why we need this. we have frontend for displaying for. All you need is function that will take new password and token, verify user and update password. we don't need this on backend part.

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,42 @@
+    const user = await userRepository.getUserById(decodedToken.userId);
+
+    if (!user) {
+      return res.status(404).send(errors.USER_NOT_FOUND);
```
we use response only in controllers. and this is service

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,42 @@
+    if (!user) {
+      return res.status(404).send(errors.USER_NOT_FOUND);
+    }
+    res.render('resetPassword', { token });
```
same

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,42 @@
+  if (!user) {
+    throw new Error(errors.USER_NOT_FOUND);
+  }
+  if (decodedToken.exp < Date.now() / 1000) {
```
I think verify will return false if token is expired. as well as if token is invalide for some other reason.

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Added authorization header for token
 * Fixed stuff from review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

AntMarkoLovric commented on app.js:
```diff
@@ -5,21 +5,21 @@ const loginRoutes = require("./routes/login")
-  origin: FRONTEND_URL
-}))
-
+app.use(cors())
```
why did you remove frontend url?

AntMarkoLovric commented on configuration/errors.js:
```diff
@@ -3,10 +3,14 @@ module.exports = {
-    
-    USER_ROLE_NOT_FOUND: "User role not found"
+    USER_ROLE_NOT_FOUND: "User role not found",
+    USER_NOT_FOUND: "User not found",
```
you added two of the same error messages, you could unite it into one and use it throughout the whole project

AntMarkoLovric commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,35 @@
+  const token = await setResetPasswordToken(user);
+
+  const templateName = "forgot-password";
+  const resetPasswordUrl = `${process.env.FRONTEND_URL}/reset-password/`;
```
extract "proces.env.FRONTEND_URL" into env.js, so you can reuse it, improves readability, and if you, at any point in development process need to change this, you will need to change in one place instead of changing in every place you use this variable

AntMarkoLovric commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,35 @@
+const env = require("../configuration/env")
+
+async function setResetPasswordToken(user) {
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
```
move this line of code into a utils/helper.js function so you can use it wherever you need token creation

AntMarkoLovric commented on services/helper.js:
```diff
@@ -0,0 +1,10 @@
+const jwt = require("jsonwebtoken");
+const env = require("../configuration/env");
+
+function generateToken(payload, expiresIn) {
```
you did this great, just move this into a utils folder

AntMarkoLovric commented on services/resetPassword.js:
```diff
@@ -0,0 +1,23 @@
+const errors = require('../configuration/errors');
+
+async function resetPassword(token, newPassword, confirmPassword) {
+  const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
```
same as the comment above regarding moving this to env file

AntMarkoLovric replied:
do we have some kind of validation if token is invalid?

TarikVelic replied:
![Screenshot_8](https://github.com/becarselman/ant-colony-pharaoh-backend/assets/101573891/b6c83281-d9c5-423b-b409-7d4333deb6e1)

we get this kind of message

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Fixing stuff from review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FIXING STUFF FROM REVIEW

--------------------------------------------------------------------
 MAIN
 * -II-
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2
--------------------------------------------------------------------
 MAIN
 * Fixing stuff from revies
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on services/forgotPassword.js:
```diff
@@ -0,0 +1,36 @@
+    throw new Error(errors.EMAIL_NOT_FOUND);
+  }
+
+  const token = await setResetPasswordToken(user);
```
we don't need extra function for this. We already have generateToken function use it here

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,24 @@
+const errors = require('../configuration/errors');
+const env = require("../configuration/env")
+
+async function resetPassword(token, newPassword, confirmPassword) {
```
we don't need to take confirm password because we will validate that on frontend.

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,24 @@
+const env = require("../configuration/env")
+
+async function resetPassword(token, newPassword, confirmPassword) {
+  const decodedToken = jwt.verify(token, env.JWT_SECRET);
```
create helper function for this

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Review fixes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Deleting confirm password in controllers
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on utils/helper.js:
```diff
@@ -0,0 +1,18 @@
+const env = require("../configuration/env");
+const errors = require('../configuration/errors');
+
+function generateToken(payload, expiresIn) {
```
please use arrow functions

becarselman commented on services/resetPassword.js:
```diff
@@ -0,0 +1,23 @@
+
+const salt = await bcrypt.genSalt(10);
+const hashedPassword = await bcrypt.hash(newPassword, salt);
+await userRepository.updateUserPassword(user._id, hashedPassword);
```
return something

TarikVelic committed:
FEATURE/TariikMVP/ForgotPasswordv2

--------------------------------------------------------------------
 MAIN
 * Small changes in code
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------
